### ISSUE TITLE
Changing a dynamically created plist's properties

### ISSUE THREAD

**Issue by amitai-hoze** [LoganWright/PlistModel#2]
Hi, great work, very useful and clever, :+1: 

I'm using a dynamically created plist, and I wanted to know what happens if I delete a property from the `PlistModel` subclass, add new ones, or add one with the same name as a previously deleted one. According to my tests, all of the deleted properties still hold their previous values, so if you delete one and return it, it will hold its previous value. Also, if you add new properties, it doesn't interfere with the old ones (e.g. reset them), which is the desired behavior for my use case. Just wanted to make sure this is indeed the regular behavior, for when I'll want to add properties to the next version of my app.

Thanks, Amitai

**Comment by LoganWright** [comment]
Hi @amitai-hoze,

Let's try a couple things:

1. Just to confirm, this is definitely a dynamic plist and there is no matching plist included in the bundle? 

2. If you're running in the simulator and rerunning the app from Xcode, sometimes the sigkill doesn't call dealloc and your values might not persisting.

3. If these don't work, would you be able to post come code?  It's probably not feasible if you're using it in a few different places, but if there's some that you have that I can use to replicate the problem, it would help me to figure out what the problem is.

Best,

Logan

**Comment by amitai-hoze** [comment]
Oh sorry, maybe my question wasn't clear enough..

There's no problem with the class, it's working great. I just wanted to find out how the class handles situations where properties were deleted or added to the `PlistModel` subclass `.h` file (that creates the plist file dynamically, like `DynamicModel` in your Xcode project).

Let's say for instance I've got this `PlistModel` subclass: 
```objective-c
@interface DynamicModel : PlistModel

@property (strong, nonatomic) NSString * name;
@property int counter;

@end
```

The code that writes to the class looks like this:
```objective-c
[DynamicModel plistNamed:@"DynamicModel"].name = someUserInputString;
[DynamicModel plistNamed:@"DynamicModel"].counter++;
```

I publish the app in the app-store, users are using it and all of them have `name` and `counter` saved values. Now I decide to add another property like so:
```objective-c
@interface DynamicModel : PlistModel

@property (strong, nonatomic) NSString * name;
@property (strong, nonatomic) NSString * newStringProperty;
@property int counter;

@end
```

are the original `name` and `counter` values guaranteed to stay the same?
And what happens if I delete the `newStringProperty` from the DynamicModel.h file, publish the app again, and then put it back and publish, will it have a `Null` value or the value saved from the last time it was saved?
In short, what is the behavior when the DynamicModel.h file changes between versions of the app (adding properties, deleting properties).
My tests indicate that all the former values are always preserved and never get reset, which is the desired behavior (in my use case at least). I just wanted to make sure I can rely on it.

Thanks,

Amitai

**Comment by LoganWright** [comment]
Ok @amitai-hoze, I understand the question much better!

So under the covers, the class is really just designating a file in the directory.  For your example above, this file is named `DynamicModel`.  When this is updated, the file is completely overwritten.  As far as I know, there isn't a way to edit only parts of the file, the way you can with something like a SQL database. I'll address some of your concerns here:

#1. Add a property that didn't exist before.

This will be nil when you load it from the old version.

#2. Remove a property that existed before.

This value will still exist in a backing dictionary when the plist is loaded from the directory.  This won't affect any of the other properties and after the first save, this will persist in the key value store until it is overwritten to nil.  If this isn't desired behavior, perhaps I should consider a good way to offer this as an option.  The reason this happens is sometimes values are accessed by key only and don't have an associated property.  You wouldn't want these to be overwritten in general.


In summation, the basic implementation is an underlying key value store, so anything added will be there until you remove it.  If you add new things, they'll be added to the rest, if you remove things there values will still exist as a key value pair until it is removed manually.

I also wanted to mention one thing about your use in the example above.  I'm not sure if this is how you're implementing it in project, but I wanted to mention it briefly.  This code:

```ObjC
[DynamicModel plistNamed:@"DynamicModel"].name = someUserInputString;
[DynamicModel plistNamed:@"DynamicModel"].counter++;
```

Would run more efficiently if it were run like this:

```ObjC
DynamicModel *dynamicModel = [DynamicModel plistNamed:@"DynamicModel"];
dynamicModel.name = someUserInputString;
dynamicModel.counter++;
```

Because it prevents loading the data from the directory unnecessarily.  Normally this isn't an extremely expensive operation, but I wanted to mention it :)

Are there any questions / issues I missed?

Best,

Logan

**Comment by amitai-hoze** [comment]
@LoganWright, thanks a lot, your answer cleared up all my concerns :relieved: 
Regarding your comment on accessing the properties more efficiently, you're right, I wrote it like that for brevity, but good comment anyways.

Thanks,
Amitai

**Comment by LoganWright** [comment]
Glad to hear it :) Good luck with the rest of your project!
